server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java

```diff
@@ -0,0 +1,191 @@
+ * <p>
+ * Each shard needs to be available and replicated in order to guarantee high availability and prevent data loses.
+ * Shards allocated on nodes scheduled for restart (using nodes shutdown API) will not degrade this indicator health.
+ */
+public class ShardsAvailabilityHealthIndicatorService implements HealthIndicatorService {
+
+    public static final String NAME = "shards_availability";
+
+    private final ClusterService clusterService;
+
+    public ShardsAvailabilityHealthIndicatorService(ClusterService clusterService) {
+        this.clusterService = clusterService;
+    }
+
+    @Override
+    public String name() {
+        return NAME;
+    }
+
+    @Override
+    public String component() {
+        return DATA;
+    }
+
+    @Override
+    public HealthIndicatorResult calculate() {
+
+        var state = clusterService.state();
+        var shutdown = state.getMetadata().custom(NodesShutdownMetadata.TYPE, NodesShutdownMetadata.EMPTY);
+        var status = new ShardAllocationStatus();
+
+        for (IndexRoutingTable indexShardRouting : state.routingTable()) {
+            for (IndexShardRoutingTable shardRouting : indexShardRouting) {
+                status.addPrimary(shardRouting.primaryShard(), shutdown);
+                for (ShardRouting replicaShard : shardRouting.replicaShards()) {
+                    status.addReplica(replicaShard, shutdown);
+                }
+            }
+        }
+
+        return createIndicator(status.getStatus(), status.getSummary(), status.getDetails());
+    }
+
+    private static class ShardAllocationCounts {
+        private int unassigned = 0;
+        private int unassigned_restarting = 0;
+        private int initializing = 0;
+        private int started = 0;
+
+        public void increment(ShardRouting routing, NodesShutdownMetadata metadata) {
+            switch (routing.state()) {
+                case UNASSIGNED -> {
+                    if (isUnassignedDueToTimelyRestart(routing, metadata)) {
+                        unassigned_restarting++;
+                    } else {
+                        unassigned++;
+                    }
+                }
+                case INITIALIZING -> initializing++;
+                case STARTED -> started++;
+            }
+        }
+    }
+
+    private static boolean isUnassignedDueToTimelyRestart(ShardRouting routing, NodesShutdownMetadata metadata) {
+        var info = routing.unassignedInfo();
+        if (info == null || info.getReason() != UnassignedInfo.Reason.NODE_RESTARTING) {
+            return false;
+        }
+        var shutdown = metadata.getAllNodeMetadataMap().get(info.getLastAllocatedNodeId());
+        if (shutdown == null || shutdown.getType() != SingleNodeShutdownMetadata.Type.RESTART) {
+            return false;
+        }
+        var now = System.currentTimeMillis();
+        var restartingAllocationDelayExpiration = shutdown.getStartedAtMillis() + shutdown.getAllocationDelay().getMillis();
```

Comment:
I just realized that it is probably better to use `info.getUnassignedTimeInMillis()` (time when shard become unassigned?) instead of `shutdown.getStartedAtMillis()` (time when shutdown command was issued).

Reply:
Fixed in the latest commit